limit_dynamic.go

```diff
@@ -23,12 +22,17 @@ type DynamicLimit struct {
 var _ Limit = (*DynamicLimit)(nil)
 
 func (l *DynamicLimit) GetMemoryLimit() int64 {
-	var mem gosigar.Mem
-	if err := mem.Get(); err != nil {
-		panic(err)
-	}
+	freemem := memory.FreeMemory()
+
+	// account for memory retained by the runtime that is actually free
+	// HeapInuse - HeapAlloc is the memory available in allocator spans
+	// HeapIdle - HeapReleased is memory held by the runtime that could be returned to the OS
+	var memstat runtime.MemStats
+	runtime.ReadMemStats(&memstat)
+
+	freemem += (memstat.HeapInuse - memstat.HeapAlloc) + (memstat.HeapIdle - memstat.HeapReleased)
```

Comment:
These are all `uint64`s. Are we 100% sure we can't underflow?

Reply:
yes, the actual address space in all 64bit processors is 48bit.

Reply:
oh you mean HeapAlloc being less than HeapInuse etc? Hrm, that _shouldn't_ happen unless the runtime goes crazy.

These subtractions are actually described in the runtime package documentation.

Reply:
https://pkg.go.dev/runtime#MemStats